README.md
## IP Address Auxiliary Platform For Dedicated Line

[![GitHub license](https://img.shields.io/github/license/thianda91/zx_fuzhu)](https://github.com/thianda91/zx_fuzhu/blob/master/LICENSE)  [![HitCount](http://hits.dwyl.io/thianda91/zx_fuzhu.svg)](http://hits.dwyl.io/thianda91/zx_fuzhu)

English | [中文](README_zh-CN.md)

### Description

Based on ThinkPHP framework.

### History

please see [CHANGELOG.md](CHANGELOG.md).

### LICENSE

MIT

CHANGELOG.md
## 更新日志

### v0.11.2(2019-09-30)

- 新增：备案信息填写 `ipbeian()` 将已有信息自动带过来。
- 修复 bug。
- 调整发送 idc/isp 备案邮件正文内容。

### v0.11.0(2019-09-25)

- 新增：管理员页面可导出 IDC/ISP 备案模板。
- 新增：`ipbeian.html` 页面提示客户经理使用非 IE 浏览器访问。
- 制作脚本最后添加空格。解决换行不生效 bug。
- 调整 `query.html` 页面上初始 `infoData` 为转义字符。
- 修复在 `todo.html` 页面上提交按钮累加的bug。

### v0.10.10(2019-09-23)

- 修复请求中含大写的情况。（全部改小写）
- 修复 `todo.html` bug，取消备案信息显示。
- 修复其他 bug。
- 调整邮件发送模板。
- 调整掩码显示方式。
- 修复发送 idc/isp 备案邮件的 bug。
- 修复首次获取验证码的逻辑 bug。
- 更新填写备案信息的模板配置。
- 修复导出工信部模板bug，修复分配 ip/vlan 后 `zxTpye` 字段丢失 bug。

### v0.10.0(2019-09-05)

- 修复 bug
- 更新前端组件
- 整理部署过程，可推广

### v0.9.0(2019-09-04)

- 重构大量功能逻辑。
- 调整工作流程，新增`ipbeian.html`及相关操作。

### v0.8.16(2019-09-03)

继续调整代码，与旧仓库分支 `v0.9.0_preview` 合并。待验证。

**todo**

- `Generateor.php` 导出模板调整。取消从前端获取模板的 header。

### v0.8.10(2019-08-29)

修改大量参数，提取到自定义配置`public/static/personalized_config.php`中，便于修改。

- 调整参数，便于推广应用。

**todo**

- 更新 3 个系统导出模板。
- 对比旧仓库的 v0.9.0 分支，合并更新。

### v0.8.3(2019-08-21)

精简项目文件，仅保留[zx_apply]模块，并重命名为[zx_fuzhu]

- 由于修改模块名称，可能存部分 bug
- 提取[zx_fuzhu]模块的文件，移除大量无用文件。
- 整合 `public/static` 内的静态文件，添加到本项目中。
- 移除部分敏感数据

#### v0.8.2(2018-10-09)

[zx_apply]模块修复 bug

- 更新状态字段 status 为
```
0： 刚申请
1： 预分配
2： IDC已备等待录资管等待工信部IP备案
3： 已录资管&工信部已备等待IDC备案
4： 等待做数据  
5： 已完成   
```
- 修复填表单时，若`neFactory`选为 ONU，对`aStation`字段的要求。（`_form_structure.html`中的`fixExtraFields()、onChangeHandler()`函数）
```
涉及的页面有
index/apply.html
index/query.html
manage/todo.html
manage/query.html
```

### v0.8.0(2018-09-25)

[zx_apply]模块添新功能

- **`todo.html`分配 IP、vlan 提交后，完美实现自动生成 idc.isp 模板数据，并做为邮件附件发送**
- 优化 idc.isp 邮件的正文内容（`Manage.php=>sendResultEmail()`）
    > 已修改为`Manage.php=>sendBeiAnResultEmail()`
- 优化`todo.html`页面提交时，冲突/错误的处理逻辑及显示

**TODO**

- IP 申请与 IDC.ISP 备案信息填写分离
- 新增客户经理填写备案信息的逻辑，填写后直接发送备案邮件。

#### v0.7.19(2018-09-21)

[zx_apply]模块修复 bug

- 修复`apply.html`选择为 ONU 后无法提交的 bug

#### v0.7.17(2018-09-10)

[zx_apply]模块修复 bug

- 更新 docs
- 修复`修改申请`报错 bug
- 修复`todo.html`分配后`vlan`无记录的 bug
- 修复其他若干 bugs

#### v0.7.12(2018-08-27)

[zx_apply]模块修复 bug

- `todo.html`和`index/query.html`编辑旧数据时自动修补新增的额外字段信息
- 修改自动修补新增的额外字段的  bugs

#### v0.7.11(2018-08-23)

[zx_apply]模块更新

- 自动生成 idc.isp 模板数据，并做为邮件附件发送
- 导出 excel 功能可选择使用前端或者后端
- 调整 LECENSE
- fix [#3](https://github.com/thianda/phpweb/issues/3), fix [#4](https://github.com/thianda/phpweb/issues/4)
- 修改若干 bugs
- `index/query.html`右键新增`额外再申请1个IP`功能
- 同 vlan 提醒后确认，可强制录入
- ONU 业务在`todo.html`可显示当前所处状态（在 vlan 字段上方）

#### v0.7.3(2018-08-05)

[zx_apply]模块更新

- 完善前端导出 excel 功能
- 分离部分`Manage.php`的操作到`Generator.php`
- `Infotables`新增 5 个额外字段（extra）
- 调整`apply.html`表单字段位置，新增字段 4 个，隐藏无需编辑的字段 4 个

### v0.7.0(2018-08-03)

[zx_apply]模块更新

- 导出 excel 功能全部迁移到前端

TODO

-  php 文件读写，发邮件带附件
-  同 vlan 强制录入

#### v0.6.7(2018-08-03)

[zx_apply]模块更新

- todo提交后发送邮件给客户经理和申请人
- close [#6](https://github.com/thianda/phpweb/issues/6)

#### v0.6.6(2018-08-01)

[zx_apply]模块更新

- 代码仓库脱敏处理，永久删除敏感信息
- 细节调整

#### v0.6.2(2018-07-27)

[zx_apply]模块更新

- 修复搜索bug [#5](https://github.com/thianda/phpweb/issues/5).

### v0.6.0(2018-07-13)

[zx_apply]模块更新

- 实现ONU业务申请时无需`A端基站`，先预分配ip不分配vlan，待做ONU上线数据时再选择`A端基站`并分配`vlan`。
- apply.html申请、query.html修改申请、todo代办页面同步修改表单动态逻辑
- 修改`Vlantables::check()`逻辑，优化存在`vlan`但不存在`aStation`时的结果

#### v0.5.5(2018-07-13)

[zx_apply]模块更新

- 调整：申请时若为ONU业务则不需要填写`A端基站`，此规则在`修改申请`时亦生效

- 修复`manage/query.html`。在此之前，批量删除时前端显示与实际不符
- 完善vlan删除与修改功能：修改状态`[0]申请`，到待办里直接修改提交即可，若不填vlan即为删除
- 修复其他若干bug

#### v0.5.1(2018-07-12)

[zx_apply]模块代码格式化

- 格式化`zx_apply`模块的代码（php、html），遵循PSR-2
- 未知原因导致部分代码丢失，已补回

### v0.5.0(2018-07-11)

[zx_apply]模块部分重构

- 前端部分重构
- 修复申请填表单的bug
- 需进一步调整

#### v0.4.18(2018-07-11)

[zx_apply]模块更新

- 更新资管导入模板
- 优化`XLSX`组件的加载逻辑
- 计划将`导出excel操作`移至前端，加速导出速度，兼容性更强
- 修复细节bug

#### v0.4.16(2018-07-05)

[zx_apply]模块更新

- 修改`index/query.html`中`afterChange`事件的bug
- 新增ip列表生成：`tool/swscript.html`

#### V0.4.13(2018-06-28)

[zx_apply]模块更新

- `manage/query.html`新增批量已完成功能，批量选取设置状态为[4]:已完成
- 导出资管模板数据时提示批量修改状态为[2]:已录资管等待IP备案

TODO：`更多操作`页面新功能：统计空闲ip

> 需基于专线类型，细节有待研究

#### V0.4.12(2018-06-26)

[zx_apply]模块修复若干bug

- 完善细节功能，修复大量bug
- 支持按照`专线类型`进行`全局搜索`以及`基本信息查询`

#### v0.4.9(2018-06-12)

- 用户可右键选择`修改申请`，修改个别字段信息后重新提交
- 用户申请提交前自动清除每个字段里的空格并给出提示

#### v0.4.8(2018-06-06)

[zx_apply]模块bug修复

- 累计优化多处UI布局
- `todo`页提示同客户名的台账信息
- `apply`页提交前过滤空格符
- 模板导出时文件名的时间字段放后面

#### v0.4.6(2018-05-29)

[zx_apply]模块更新

- 视图显示优化、安全事件逻辑优化
- 显示和导出全量数据可基于专线类别筛选，申请与待办暂未实现
- 查询、导出操作均记录log

#### V0.4.5(2018-05-24)

[zx_apply]模块bug修复

- 修复ip分配时的小bug
- 调整查询页面排序为`status,create_time desc`
- 更新资管系统导出模板
- 调整各种模板导出时文件名的命名规则

#### v0.4.4(2018-05-18)

[zx_apply]模块大量更新。

- 待办页`vlan`字段不再是必填
- 待办页“自动预分配”结果显示由`alert`改为`window`
- 查询页显示详细信息布局优化
- 查询页取消显示备注列
- 调整待办通知的邮件内容
- 申请人查询页可查询历史数据的基本信息（去除客户联系方式等敏感信息，并限制查询次数）

#### v0.4.3(2018-05-16)

- [zx_apply]模块预分配ip后给申请人自动发送邮件提醒（手动判断是否发送）

#### v0.4.2(2018-05-11)

#### v0.4.1(2018-05-04)

[zx_apply]模块开始公测使用

- 添加`jsSheet`组件，用于zx_apply前端导出全量台账的xlsx文件
- 敏感配置信息独立存储

### v0.4.0(2018-04-02)

[zx_apply]模块大量更新。

- `Infotables`增加 ip掩码和ipB掩码，互联网ip掩码默认32，其他默认30，所有ipB掩码默认29
- 取消`Iptables`数据表，保留其model类及方法
- 定位为辅助平台，提供有限的台账维护功能。主要功能为限制数据录入的规范性。导出为各项工作所需的数据。包含但不限于：
  - 台账格式
  - 资管流程模板格式
  - 做数据脚本
  - ip备案模板格式（2种）
- 管理员查询页面可右键调用数据制作脚本生成、资管系统录入模板、集团IP备案导入模板、工信部IP备案导入模板功能。
- 更新`handsontable`组件到v1.18.1/v0.38.1

### v0.3.2(2018-01-31)

- 更新`dhtmlx`组件到5.1.0
- 修复esserver的bug
- 添加系统log记录功能
- 添加zx_apply模块的登陆控制


application/zx_fuzhu/controller/Index.php
<?php

namespace app\zx_fuzhu\controller;

use think\Db;
use think\Cache;
use app\zx_fuzhu\model\Infotables;
use app\zx_fuzhu\model\Vlantables;


class Index extends Common
{
	public function ch2arr($str)
	// 未使用
	{
		$length = mb_strlen($str, 'utf-8');
		$array = [];
		for ($i = 0; $i < $length; $i++)
			$array[] = mb_substr($str, $i, 1, 'utf-8');
		return $array;
	}

	/**
	 * 首页-登录
	 *
	 * @return mixed|string|void
	 */
	public function index()
	{
		if (request()->isGet()) {
			if (request()->pathinfo() != request()->module() . "/index/index.html") {
				return $this->redirect("index");
			}
			return $this->fetch();
		}
		if (request()->isPost()) {
			// post请求 验证登陆
			$user = Db::table("phpweb_check")->where('email', input("post.email"))->order("time desc")->find();
			$msg = "";
			if (!$user) {
				return $this->result(["code" => 0], 0, "该邮箱还未申请验证码");
			}
			if ($user["code"] != input("post.code")) {
				$msg = "验证码错误";
			} else {
				// 验证码正确，继续验证申请人姓名
				if ($user["name"] != input("post.name")) {
					$msg = "申请人姓名与申请时不一致<br />申请时为：<b>" . $user["name"] . "</b><br />申请时间：" . $user["time"];
				}
			}
			if ($msg) {
				$this->writeLog("登陆", "failed", $msg);
				return $this->error($msg, null, input("post."));
			} else if (time() - strtotime($user["time"]) > 3600 * 24 * 15) { // 15天内可直接登陆
				$msg = "登陆超时，请重新获取验证码。";
				$this->writeLog("登陆", "failed", $msg);
				unset($user["code"]);
				return $this->error($msg, "index", $user);
			} else {
				$e = explode("@", $user["email"]);
				// 附加role。
				if ($e[1] == "ln.chinamobile.com" && in_array($e[0], config("manageEmails"))) {
					$user["role"] = "manage";
				} else {
					$user["role"] = "index";
				}
				session("user", $user);
				$msg = "欢迎回来，" . $user["name"] . "。";
				$this->writeLog("登陆", "success", $msg);
				$hash = input("post.hash");
				$hashToURL = request()->pathinfo() == request()->module() . "/index/index.html" ? $hash != null ? $hash != "logout" ? "/" . request()->module() . "/" . $hash . ".html" : null : null : null;
				$to_url = session("to_url") ? session("to_url") : session("user.role") . "/query";
				// 跳转优先级：1. location.hash、2. session("to_url")、3. session("user.role") . "/query"
				$url =  $hashToURL ? $hashToURL : $to_url;
				session("to_url", null);
				return $this->success($msg, $url, $user);
			}
		}
	}

	protected function writeLog($k = '', $status = '', $msg = '')
	{
		$this->log($k, [
			"status" => $status,
			"name" => input("post.name"),
			"email" => input("post.email"),
			"msg" => strip_tags($msg)
		]);
	}

	/**
	 * 数据专线申请开通
	 */
	public function apply()
	{
		if (request()->isGet()) {
			return $this->fetch("index/apply");
		} else if (request()->isPost()) {
			$data = input("post.");
			$this->checkInstanceID(null, $data["instanceId"]); // 检查instanceId
			// 工程申请时不再填写idc/isp备案信息，下面的转换不再需要。
			// $extraHeader = config("extraInfo");
			// foreach ($extraHeader as $k => $v) {
			// 	$data["extra"][$v] = $data[$v];
			// 	unset($data[$v]);
			// }
			$result = Infotables::createInfo($data, "apply");
			// 发邮件通知 给客户经理，抄送 IP 地址管理员、当前申请人
			$subject = "[待办]ip申请-" . ($data["ifOnu"] ? "onu" : "9312") . "-" . $data["cName"] . "-" . $data["instanceId"];
			$body = $this->todo_link_str();
			$this->sendManageNotice($subject, $body, true);
			$v = [
				"username" => session("user.name"),
				"email" => session("user.email"),
				"cName" => $data["cName"],
				"instanceId" => $data["instanceId"]
			];
			$this->log("提交申请", $v);
			$redirectUrl = "../" . session("user.role") . "/query.html";
			return $this->result(null, $result, $redirectUrl);
		}
	}

	/**
	 * 修改提交
	 */
	public function re_apply()
	{
		if (request()->isGet()) {
			return $this->error("Nothing here. Do not try again!");
		}
		$data = input("post.");
		$oldData = Infotables::get(input("param.old"))->toArray();
		$updates = []; // 记录有更新的字段
		$change = '';
		/* 不验证实例标识重复与否 */
		$extraHeader = config("extraInfo");
		foreach ($extraHeader as $k => $v) {
			$data["extra"][$v] = $data[$v];
			unset($data[$v]);
			if ($data["extra"][$v] != $oldData["extra"][$v]) {
				$change .= $k . " => [" . $oldData["extra"][$v] . "] 改为 [" . $data["extra"][$v] . "]\r\n";
				$updates[$k] = "[" . $oldData["extra"][$v] . "]=>[" . $data["extra"][$v] . "]";
			}
		}
		// 不验证 instanceID
		$result = Infotables::createInfo($data, "apply");
		$this->queryDelete(["id" => [input("param.old")]]);
		// 发邮件通知
		$subject = "[待办]修改申请-" . $data["aPerson"] . "-" . $data["cName"];
		$oldInfo = "<pre style='color:#088'>A端基站： " . $oldData["aStation"] . "\r\nip: " . $oldData["ip"] . "\r\nvlan: " . $oldData["vlan"] . "</pre>";
		$body = "<p>原分配信息：</p>" . $oldInfo;
		$body .= "<p>修改内容：</p><pre style='color:#088bff'>";
		foreach ($data as $k => $v) {
			if ($k = 'extra') {
				continue;
			}
			// last change from v0.7.17
			if ($v != $oldData[$k]) {
				$change .= $k . ' => [' . $oldData[$k] . '] 改为 [' . $v . "]\r\n";
				$updates[$k] = '[' . $oldData[$k] . ']=>[' . $v . ']';
			}
		}
		$change == '' && $change = "无";
		$body .= $change;
		$body .= $this->todo_link_str();
		$this->sendManageNotice($subject, $body, true);
		$v = [
			"username" => session("user.name"),
			"email" => session("user.email"),
			"cName" => $data["cName"],
			"instanceId" => $data["instanceId"],
			"oldInfo" => $oldInfo,
			"updates" => $updates
		];
		$this->log("修改申请", $v);
		$redirectUrl = "../" . session("user.role") . "/query.html";
		return $this->result(null, $result, $redirectUrl);
	}

	/**
	 * IP 备案信息填写
	 *
	 * @return void
	 */
	public function ipbeian()
	{
		if (request()->isGet()) {
			return $this->fetch();
		}
		if (request()->isPost()) {
			if (!input("?post.cName")) {
				return $this->result("没有传参", 0);;
			}
			// 提交时更新 Infotables，并发送idc备案邮件
			$data = input('post.');
			$id = $data['cName'];
			unset($data['cName']);
			$extraHeader = config("extraInfo");
			foreach ($extraHeader as $k => $v) {
				if (isset($data[$v])) {
					$data["extra"][$v] = $data[$v];
					unset($data[$v]);
				}
			}
			$infotables = new Infotables();
			$data['status'] = 2;
			$result = $infotables->isUpdate(true)->allowField(true)->save($data, ["id" => $id]);
			// 自动发IDC备案信息给厂家
			return $this->sendBeiAnResultEmail($id);
		}
		if (request()->isPut()) {
			if (input("param.r") == "get_cnames") {
				$result = Infotables::where('mEmail', session('user.email'))->where('status', '<', 3)->order('id desc')->column('id,cName');
				return $result;
			}
			if (input("param.r") == "get_detail" && input("?param.id")) {
				$order_en = '6,12,13,14,28';
				$field = $this->getHeader("zx_apply-new-rb", $order_en, true);
				$result = Infotables::field($field)->select(input("param.id"))[0];
				return $result;
			}
		}
	}
	public function tt()
	{
		return dump($this->getHeader("zx_apply-new-rb", '6,10,3,39,40,32,33,38,7,43,23'));
	}
	/**
	 * 发送 IDC 备案信息给厂家联系人
	 *
	 * @param string $id
	 * @return void
	 */
	public function sendBeiAnResultEmail($id = '')
	{
		$order_en = "6,10,3,15,17,23,28,7";
		// cName,ip,bandWidth,mPerson,mEmail,aEmail,extra,cAddress
		$order_cn = "6,10,3,39,40,33,38,7,43,23";
		$field = $this->getHeader("zx_apply-new-rb", $order_en, true);
		// 获取表格标题
		$header_cn = explode(",", $this->getHeader("zx_apply-new-rb", $order_cn));
		$header_en = explode(",", $this->getHeader("zx_apply-new-rb", $order_cn, true));
		// 获取表格内容
		$db = Infotables::field($field)->find($id)->toArray();
		// 转换表格内容的 extra 信息
		foreach ($db["extra"] as $k => $v) {
			$db[$k] = $v;
		}
		unset($db["extra"]);
		//$values = array_values($db);
		$title = config('idc.title_city') . 'IDC.ISP-' . $db['cName'] . "-" . $db['ip'];
		$contact = config('idc_contact');
		$contact_str = implode(',', $contact);
		$body = "--本邮件用来idc备案--";
		$body .= "<br><table style='border-collapse:collapse;border:none;'>";
		for ($i = 0; $i < count($header_cn); $i++) {
			$body .= "<tr>";
			$body .= "<th style='width:200px;border:solid #666 1px;'>" . $header_cn[$i] . "</th><td style='width:500px;border:solid #666 1px;'>" . $db[$header_en[$i]] . "</td>";
			$body .= "</tr>";
		}
		$body .= "</table>";
		$body .= "<p style='color:#000;background-color:#ccc;font-size:12px;'>备案信息由客户经理<span style='color:blue'>" . $db['mPerson'] . "(" . $db['mEmail'];
		$body .= ")</span>负责填写，并确保其完整性、准确性。<br>本邮件由辅助平台自动发送给厂家联系人：" . $contact_str . "，并抄送给 IP 地址管理员。</p>";
		$body .= "如备案信息有误，请厂家联系人联系客户经理。由客户经理负责登陆辅助平台进行更正重新提交，或直接回复给厂家联系人。";
		$address = $contact;
		$address['CCm'] = $db['mEmail'];
		$address['CCa'] = $db['aEmail'];
		foreach (config("manageEmails") as $k => $v) {
			$address['CC' . $k] = $v . "@ln.chinamobile.com";
		}
		$attachment = Generator::generateIDCinfoFiles($id);
		$result = $this->sendEmail($address, $title, $body, $attachment);
		unlink($attachment);
		return $this->result($result, is_bool($result) ? 1 : 0);
	}

	/**
	 * 根据label、order 获取表格的 header!
	 * $v为false，获取option(default)；为ture，获取value
	 *
	 * @param String $label            
	 * @param String $order            
	 * @param boolean $v            
	 * @return string
	 */
	public function getHeader($label = "label", $order = "order", $v = false)
	{
		if ($label === "label" || $order === "order") {
			return "{msg:\"你要搞什么？\"}"; // 未输入参数label或order
		}
		$_headerData = Db::table("phpweb_sysinfo")->field("value,option")->where(["label" => $label])->order("id")->select();
		$orderArr = explode(",", $order);
		$headerArr = [];
		$sub = $v ? "value" : "option";
		foreach ($orderArr as $o) {
			$headerArr[] = $_headerData[$o][$sub];
		}
		return implode(",", $headerArr);
	}

	/**
	 * 根据order获取handsontable组件的colWidth
	 *
	 * @param unknown $order            
	 * @return string|void
	 */
	protected function getColWidths($order = null)
	{
		if (!is_null($order)) {
			$orderArr = explode(",", $order);
			$result = [];
			foreach ($orderArr as $v) {
				$result[] = config("colWidth")[$v];
			}
			return implode(",", $result);
		}
		return $this->result(null, 0, "~缺参数~");
	}

	/**
	 * 信息查询
	 */
	public function query()
	{
		if (request()->isGet()) {
			// 访问
			$aStation = array_keys(config("aStation"));
			$zxTitle = [
				"label" => "zx_apply-new-rb",
				"order" => "24,1,2,3,4,5,6,7,8,9,10,12,13,14,15,16,17,18,19,20,29,30,31,32,33,34,35,36,37,38,39,40,41,42,22,23,26"
			];
			$this->assign([
				"aStationData" => implode(",", $aStation),
				"colHeaderData" => $this->getHeader($zxTitle["label"], $zxTitle["order"]),
				"colWidthsData" => $this->getColWidths($zxTitle["order"]),
				"data" => addslashes($this->getInfoData()->toJson()),
			]);
			return $this->fetch();
		}
		if (request()->isPost()) {
			// 获取台账
			input("post.r") == "info" && $data = $this->getInfoData(input("post.zxType"))->toArray();
			input("post.r") == "search" && $data = $this->querySearch(input("post."));
			input("post.r") == "brief" && $data = $this->querySearchBrief(input("post."));
			input("get.r") == "update" && $data = $this->queryUpdateInfo(input("post."));
			return $data;
		}
		if (request()->isPut()) {
			// 相关操作
			input("post.r") == "copy_one_more" && $data = $this->queryCopyOne(input("post."));
			input("post.r") == "export" && $data = $this->queryExport();
			return $data;
		}
	}

	/**
	 * 获取台账信息
	 *
	 * @param string $zxType            
	 * @param number $limit            
	 * @return \think\model\Collection|\think\Collection
	 */
	protected function getInfoData($zxType = "互联网", $limit = 100)
	{
		$where = "";
		if (session("user.role") != "manage") {
			$where = ["aEmail" => session("user.email")];
		}
		return collection(Infotables::where("zxType", $zxType)->where($where)->order("status,ip desc,create_time desc")->limit($limit)->select());
	}

	/**
	 * 全局查询
	 *
	 * @param unknown $data            
	 * @return array
	 */
	protected function querySearch($data)
	{
		$where = "";
		if (session("user.role") != "manage") {
			$where = ["aEmail" => session("user.email")];
		}
		return collection(Infotables::where("zxType", $data["zxType"])->where($where)->where($data["where"][0], "like", "%" . $data["where"][2] . "%")->order("ip desc")->select())->toArray();
	}

	/**
	 * 基本信息查询
	 *
	 * @param unknown $data            
	 * @return array
	 */
	private function querySearchBrief($data)
	{
		if (!isset($data["where"][2]) || $data["where"][2] == "") {
			return;
		}
		$field = "create_time,instanceId,cName,cAddress,vlan,ip,aPerson,aEmail";
		$result = collection(Infotables::where("zxType", $data["zxType"])->where($data["where"][0], "like", "%" . $data["where"][2] . "%")->field($field)->order("ip desc")->select())->toArray();
		$logValue  = $data;
		$logValue["resultLen"] = count($result);
		$logValue["user"] = session("user.name");
		$logValue["url"] = request()->url(true);
		$this->log("基本信息查询", $logValue);
		if (Cache::get('querySearchBriefTimes')) {
			Cache::inc('querySearchBriefTimes');
		} else {
			Cache::set('querySearchBriefTimes', 1, 600);
		}
		$querySearchBriefTimes = Cache::get('querySearchBriefTimes');
		if ($querySearchBriefTimes > 12) {
			$address = config("manageEmails");
			foreach ($address as $k => $v) {
				$address[$k] = $v . "@ln.chinamobile.com";
			}
			if ($querySearchBriefTimes > 14) {
				$msg = session("user.name") . "已被系统强制登出，原因：查询频繁，单位时间内累计" . $querySearchBriefTimes . "次";
				$this->log("强制登出", $msg);
				$this->noticeManage("[频繁查询]" . session("user.name") . "-10分钟内：" . $querySearchBriefTimes, null, $address);
				session(null);
				return $this->error("已退出登陆，请勿频繁查询！", "index");
			}
			return $querySearchBriefTimes;
		}
		return $result;
	}

	/**
	 * 从query.html更新台账
	 *
	 * @param unknown $updateData            
	 * @return number|\think\false
	 */
	protected function queryUpdateInfo($updateData)
	{
		$result = 0;
		$dbNew = [];
		foreach ($updateData as $k => $v) {
			$infotables = new Infotables();
			$line_and_id = explode("-", $k);
			$extraHeader = config("extraInfo");
			foreach ($extraHeader as $e => $exH) {
				if (isset($v[$exH])) { // extra 部分有更新，需要先获取完整的
					$v["extra"][$exH] = $v[$exH];
					unset($v[$exH]);
				}
			}
			$result += $infotables->isUpdate(true)->allowField(true)->save($v, ["id" => $line_and_id[1]]);
			// 反查询刚才修改后的数据库里的值，用于前后端数据的一致性
			$data = $infotables->where("id", $line_and_id[1])->find();
			foreach ($v as $kk => $vv) {
				$dbNew[$k][$kk] = $data->$kk;
			}
			$infotables = null;
		}
		$logValue = [
			"sqlResult" => $result,
			"name" => session("user.name"),
			"email" => session("user.email"),
			"updateData" => $updateData,
		];
		$this->log("台账更新", $logValue);
		return $this->result($dbNew, 1, $result);
	}

	/**
	 * 从query.html删除台账条目
	 *
	 * @param unknown $input            
	 */
	protected function queryDelete($input)
	{
		$result = Infotables::destroy($input["id"]);
		// 操作人记录到备注里
		Infotables::where('id', 2684)->exp('marks', 'concat(marks,\'' . session("user.name") . "已删;" . "')")->update();
		// 同步删除vlantables
		foreach ($input["id"] as $id) {
			Vlantables::destroy(["infoId" => $id]);
		}
		$logValue = [
			"sqlResult" => $result,
			"name" => session("user.name"),
			"email" => session("user.email"),
			"deleteData" => $input,
		];
		$this->log("台账更新", $logValue);
		return $result;
	}
	/**
	 * 额外再申请 IP 操作
	 *
	 * @return void
	 */
	protected function queryCopyOne($input)
	{
		if (request()->isGet()) {
			return $this->error("Nothing here. Do not try again!");
		}
		$id = $input["copyOne"];
		/* 不验证实例标识重复与否 */
		$data = Infotables::get($id)->getData();
		// json str 转 object
		$data['extra'] = json_decode($data['extra']);
		// 获取原始数据，去除 id，直接保存
		unset($data["id"]);
		$result = Infotables::createInfo($data, "apply");
		// 发邮件通知
		$subject = "[待办]额外申请-" . $data["aPerson"] . "-" . $data["cName"];
		$oldInfo = "<pre>A端基站： " . $data["aStation"] . "\r\nip: " . long2ip($data["ip"]) . "\r\nvlan: " . $data["vlan"] . "</pre>";
		$body = "<p>原分配信息：</p>" . $oldInfo;
		$body .= "<p>现申请额外的IP，需要您核实。";
		$body .= $this->todo_link_str();
		$this->sendManageNotice($subject, $body);
		$logValue = [
			"username" => session("user.name"),
			"email" => session("user.email"),
			"cName" => $data["cName"],
			"instanceId" => $data["instanceId"],
			"oldInfo" => $oldInfo
		];
		$this->log("额外申请", $logValue);
		$redirectUrl = "../" . session("user.role") . "/query.html";
		return $this->result(null, $result, $redirectUrl);
	}
	/**
	 * 导出全量台账-基于专线类型
	 *
	 * @param string $zxType            
	 * @return string[]|array[]
	 */
	protected function queryExport($zxType = "互联网")
	{
		if ($zxType == "互联网") {
			$colHeader = "申请时间,产品实例标识,带宽,网元厂家,A端基站,客户名称,单位详细地址,客户需求说明(选填),VLAN,IP,联系人姓名(客户侧),联系电话(客户侧),联系人邮箱(客户侧)*,负责人姓名(移动侧)*,负责人电话(移动侧)*,负责人邮箱(移动侧)*,备注,是否ONU带\n(默认为否),单位性质*,单位分类*,行业分类*,使用单位证件类型*,使用单位证件号码*,单位所在省*,单位所在市*,单位所在县*,应用服务类型*,单位属性,网络安全责任人,责任人身份证号,责任人电话,责任人邮箱";
			$colName = "create_time,instanceId,bandWidth,neFactory,aStation,cName,cAddress,cNeeds,vlan,ip,cPerson,cPhone,cEmail,mPerson,mPhone,mEmail,marks,ifOnu,extra.unitProperty,extra.unitCategory,extra.industryCategory,extra.credential,extra.credentialnum,extra.province,extra.city,extra.county,extra.appServType,extra.unitAttribute,extra.securityPerson,extra.securityPersonID,extra.securityPhone,extra.securityEmail";
			$field = "";
		} else if ($zxType == "营业厅") {
			$colHeader = "申请时间,产品实例标识,网元厂家,A端基站,客户名称,单位详细地址,VLAN,互联IP,业务IP,联系人姓名(客户侧),联系电话(客户侧)";
			$colName = "create_time,instanceId,neFactory,aStation,cName,cAddress,,vlan,ip,ipB,cPerson,cPhone";
			$field = $colName;
		} else if ($zxType == "卫生网") {
			$colHeader = "申请时间,产品实例标识,网元厂家,A端基站,客户名称,单位详细地址,VLAN,互联IP,业务IP,联系人姓名(客户侧),联系电话(客户侧)";
			$colName = "create_time,instanceId,neFactory,aStation,cName,cAddress,vlan,ip,ipB,cPerson,cPhone";
			$field = $colName;
		} else if ($zxType == "平安校园") {
			$colHeader = "申请时间,产品实例标识,客户名称,单位详细地址,VLAN,监控IP";
			$colName = "create_time,instanceId,cName,cAddress,vlan,ip";
			$field = $colName;
		}
		if (session("user.role") == "manage") {
			$data = collection(Infotables::field($field)->where("zxType", $zxType)->order("ip")->select())->toArray();
		} else {
			$data = collection(Infotables::field($field)->where("aPerson", session("user.name"))->order("ip")->select())->toArray();
		}
		$logValue = [
			"dataNum" => count($data),
			"zxType" => $zxType,
			"username" => session("user.name"),
			"email" => session("user.email")
		];
		$this->log("导出全量数据", $logValue);
		return [
			"data" => $data,
			"colHeader" => $colHeader,
			"colName" => $colName
		];
	}

	/**
	 * 检查 instanceId 是否重复
	 * 可输入$data数组或instanceId
	 *
	 * @param unknown $info            
	 * @param unknown $data            
	 */
	protected function checkInstanceID($info, $data)
	{
		if (null == $info) {
			$instanceId = $data;
		} else if ($info["id"] != $data["id"]) {
			$instanceId = $data["instanceId"];
		} else {
			return;
		}
		$info = Infotables::get([
			"instanceId" => $instanceId
		]);
		if ($info) {
			return $this->error("实例标识重复，请重试", null, "该实例标识对应客户名为：<br>" . $info["cName"] . "<br>申请人：" . $info["aPerson"]);
		}
	}

	/**
	 * 给管理员发送通知, 可选：抄送给当前用户
	 *
	 * @param string $subject
	 * @param string $body
	 * @param boolean $addCurrentUser
	 */
	protected function sendManageNotice($subject = '', $body = '', $addCurrentUser = false)
	{
		$address = config("manageEmails");
		foreach ($address as $k => $v) {
			$address[$k] = $v . "@ln.chinamobile.com";
		}
		$addCurrentUser && $address["CC1"] = session("user.email");
		$this->sendEmail($address, $subject, $body);
	}

	/**
	 * 生成发送代办邮件时的访问链接
	 *
	 * @param string $hash
	 */
	protected function todo_link_str($hash = 'manage/todo')
	{
		return "<p>请登陆系统及时处理：</p><br> 内网： <a href='http://" . config('address_local') . "/" . config('moduleName') . "/index/index.html#" . $hash . "'>http://" . config('address_local') . "/" . config('moduleName') . "/index/index.html#" . $hash . "</a><br>外网： <a href='https://" . config('address_wide') . "/" . config('moduleName') . "/index/index.html#" . $hash . "'>https://" . config('address_wide') . "/" . config('moduleName') . "/index/index.html#" . $hash . "</a>";
	}
}
